matrix/app.py

```diff
@@ -1,139 +1,69 @@
+
+        self.ui = WebUI(
+            on_rotation_clockwise=self.handle_rotation_clockwise,
+            on_rotation_counterclockwise=self.handle_rotation_counterclockwise,
+            on_press=self.handle_press,
         )
 
-    def get_screen(self, index: int) -> Image:
-        if all(screen.image is None for screen in self.screens):
-            return get_image_no_connection()
-        screens = [screen.image for screen in self.screens if screen.image is not None]
-        return screens[index % len(screens)]
+        self.hardware.dial.when_rotated_clockwise = self.handle_rotation_clockwise
+        self.hardware.dial.when_rotated_counter_clockwise = (
+            self.handle_rotation_counterclockwise
+        )
+        self.hardware.button.when_pressed = self.handle_press
 
-    def handle_rotation_clockwise(self):
-        if self.in_menu:
-            self.menu.handle_rotation_clockwise()
-        else:
-            self.screen_index += 1
-            self.next_refresh_time = time.time() + 10
+        self.signal_update = threading.Event()
 
-    def handle_rotation_counter_clockwise(self):
-        if self.in_menu:
-            self.menu.handle_rotation_counter_clockwise()
-        else:
-            self.screen_index -= 1
-            self.next_refresh_time = time.time() + 10
+    def change_mode(self, mode: ModeType) -> None:
+        self.active_mode = mode
 
-    def rotate_screen(self) -> Image:
-        while True:
-            self.screen_index += 1
-            self.next_refresh_time = time.time() + 5
+    def handle_rotation_clockwise(self):
+        self.modes[self.active_mode].handle_encoder_clockwise()
+        self.signal_update.set()
 
-            while time.time() < self.next_refresh_time:
-                # another thread may modify next_refresh_time
-                time.sleep(0.1)
+    def handle_rotation_counterclockwise(self):
+        self.modes[self.active_mode].handle_encoder_counterclockwise()
+        self.signal_update.set()
 
-    def wait_for_button(self, timeout: float = 0.05):
-        self.button.wait_for_active(timeout=timeout)
-        if self.button.is_active:
-            self.button.wait_for_inactive()
-            self.in_menu = True
+    def handle_press(self):
+        self.modes[self.active_mode].handle_encoder_push()
+        self.signal_update.set()
 
-    def __next__(self) -> Image:
+    def run(self):
         try:
-            if self.in_menu:
-                exit_menu = self.menu.wait_for_button()
-                if exit_menu:
-                    self.in_menu = False
-                else:
-                    return self.menu.draw()
-
-            self.wait_for_button()
-            if is_eleven_eleven():
-                return get_image_fish()
-
-            return self.get_screen(self.screen_index)
-        except Exception:
-            self.log.exception("Exception during rendering")
-            return get_image_no_connection()
```

Comment:
i think my changes broke the no connection / error image and retry logic. need to fix